File: src/lib/mime.ts
/**
 * Extension → content-type mapping for the static raw endpoints (repo blobs and note files).
 *
 * A static host decides the real `Content-Type` at serve time from the file extension, so this
 * table only governs the response Astro builds — what `astro preview` and the e2e server hand
 * back. It exists as one shared module rather than a copy per endpoint because the two must
 * never disagree about, say, whether `.svg` is served as an image.
 *
 * Pure and dependency-free: no node imports, so it stays usable from anywhere.
 */

/**
 * Lowercased extension (no dot) → content type. Small on purpose: it covers what actually
 * turns up in a source repo or a notes folder, and everything else falls back to plain text.
 */
const MIME: Record<string, string> = {
  png: 'image/png',
  jpg: 'image/jpeg',
  jpeg: 'image/jpeg',
  gif: 'image/gif',
  webp: 'image/webp',
  svg: 'image/svg+xml',
  ico: 'image/x-icon',
  json: 'application/json',
  pdf: 'application/pdf',
  zip: 'application/zip',
  wasm: 'application/wasm',
  html: 'text/html; charset=utf-8',
  htm: 'text/html; charset=utf-8',
  css: 'text/css; charset=utf-8',
  js: 'text/javascript; charset=utf-8',
  mjs: 'text/javascript; charset=utf-8',
  woff: 'font/woff',
  woff2: 'font/woff2',
  ttf: 'font/ttf',
  mp3: 'audio/mpeg',
  mp4: 'video/mp4',
  webm: 'video/webm',
};

/**
 * Content type for a file path.
 *
 * Defaults to UTF-8 plain text: an unknown extension in a source repo is far more likely to be
 * text than a binary the browser should download, and serving it as `application/octet-stream`
 * would turn "view the raw file" into a download prompt.
 *
 * @param path File path or name; only the final extension is inspected.
 * @returns A `Content-Type` header value.
 */
export function mimeFor(path: string): string {
  const dot = path.lastIndexOf('.');
  const ext = dot === -1 ? '' : path.slice(dot + 1).toLowerCase();
  return MIME[ext] ?? 'text/plain; charset=utf-8';
}

File: internal/serve/mime.go
package serve

import (
	"path/filepath"
	"strings"
)

// mimeByExt is the port of src/lib/mime.ts, extension (no dot, lowercase) → content type.
//
// It is a fixed table rather than mime.TypeByExtension on purpose. Go's stdlib lookup consults
// the machine — the Windows registry, /etc/mime.types — so the same file could be served as
// text/plain here and application/octet-stream on the next box, and the raw endpoints exist to
// hand back committed bytes with a type a browser will display rather than download.
//
// Small on purpose too: it covers what actually turns up in a source repository or a notes
// folder, and everything else is text.
var mimeByExt = map[string]string{
	"png":   "image/png",
	"jpg":   "image/jpeg",
	"jpeg":  "image/jpeg",
	"gif":   "image/gif",
	"webp":  "image/webp",
	"svg":   "image/svg+xml",
	"ico":   "image/x-icon",
	"json":  "application/json",
	"pdf":   "application/pdf",
	"zip":   "application/zip",
	"wasm":  "application/wasm",
	"html":  "text/html; charset=utf-8",
	"htm":   "text/html; charset=utf-8",
	"css":   "text/css; charset=utf-8",
	"js":    "text/javascript; charset=utf-8",
	"mjs":   "text/javascript; charset=utf-8",
	"woff":  "font/woff",
	"woff2": "font/woff2",
	"ttf":   "font/ttf",
	"mp3":   "audio/mpeg",
	"mp4":   "video/mp4",
	"webm":  "video/webm",
}

// MimeFor returns the Content-Type for a file path; only the final extension is inspected.
//
// Unknown falls back to UTF-8 plain text, which is the interesting half of the decision: an
// unrecognised extension in a source repository is far more likely to be text than a binary,
// and application/octet-stream would turn "view the raw file" into a download prompt.
func MimeFor(name string) string {
	ext := strings.ToLower(strings.TrimPrefix(filepath.Ext(name), "."))
	if t, ok := mimeByExt[ext]; ok {
		return t
	}
	return "text/plain; charset=utf-8"
}
